File: src/rac/services/diff.py
"""AST-level requirement diff between two versions of an artifact.

``diff`` compares two :class:`~rac.core.models.Product` ASTs and classifies
what changed. It reads only the parsed structure — requirements, success
metrics, risks — never the raw Markdown, so cosmetic edits (reflowed prose,
reordered paragraphs) never register as a change.

Matching rules:

- Requirements are keyed by their ``[REQ-NNN]`` id. A shared id with identical
  text is unchanged (omitted); a shared id with different text is *modified*;
  an id present only in the new version is *added*; one present only in the old
  is *removed*.
- Success metrics and risks are compared as plain strings by exact equality:
  each direction's delta is the ordered set difference.
"""

from __future__ import annotations

from rac.core.models import Diff, Product, Requirement, RequirementChange


def _index_by_id(requirements: list[Requirement]) -> dict[str, Requirement]:
    # A duplicate id is a validation error handled elsewhere; here the last
    # occurrence wins so the mapping — and thus the diff — stays well-defined.
    return {requirement.id: requirement for requirement in requirements}


def _ordered_difference(source: list[str], other: list[str]) -> list[str]:
    """Items in ``source`` absent from ``other``, in ``source`` order, de-duped."""
    excluded = set(other)
    seen: set[str] = set()
    result: list[str] = []
    for item in source:
        if item not in excluded and item not in seen:
            seen.add(item)
            result.append(item)
    return result


def diff(old: Product, new: Product) -> Diff:
    """Return the classified :class:`Diff` between ``old`` and ``new``."""
    old_by_id = _index_by_id(old.requirements)
    new_by_id = _index_by_id(new.requirements)

    result = Diff()

    # Added and modified are emitted in new-file order (first appearance of each id).
    for req_id, new_req in new_by_id.items():
        old_req = old_by_id.get(req_id)
        if old_req is None:
            result.added_requirements.append(new_req)
        elif old_req.text != new_req.text:
            result.modified_requirements.append(
                RequirementChange(id=req_id, old_text=old_req.text, new_text=new_req.text)
            )

    # Removed are emitted in old-file order.
    for req_id, old_req in old_by_id.items():
        if req_id not in new_by_id:
            result.removed_requirements.append(old_req)

    result.added_metrics = _ordered_difference(new.success_metrics, old.success_metrics)
    result.removed_metrics = _ordered_difference(old.success_metrics, new.success_metrics)
    result.added_risks = _ordered_difference(new.risks, old.risks)
    result.removed_risks = _ordered_difference(old.risks, new.risks)

    return result

File: rust/rac-engine/src/diff.rs
//! AST diff (`decided.services.diff`): compare two parsed products and classify
//! the changes. Pure AST work — no git, no revisions, no raw-text diffing.
//!
//! - Requirements match by ID: same ID + same text -> unchanged (omitted);
//!   same ID, different text -> modified; ID only in the new -> added; ID
//!   only in the old -> removed.
//! - Metrics and risks are ordered set-difference, de-duped, preserving
//!   source order.

use crate::markdown::Requirement;
use crate::parse::Artifact;

/// `RequirementChange` — a requirement whose text changed (same ID). Field
/// order mirrors the Python dataclass (`id, old_text, new_text`).
#[derive(Debug, Clone)]
pub struct RequirementChange {
    pub id: String,
    pub old_text: String,
    pub new_text: String,
}

/// `Diff` — the classified differences between two products.
#[derive(Debug, Clone, Default)]
pub struct Diff {
    pub added_requirements: Vec<Requirement>,
    pub removed_requirements: Vec<Requirement>,
    pub modified_requirements: Vec<RequirementChange>,
    pub added_metrics: Vec<String>,
    pub removed_metrics: Vec<String>,
    pub added_risks: Vec<String>,
    pub removed_risks: Vec<String>,
}

impl Diff {
    /// True when nothing changed across any comparison unit.
    pub fn is_empty(&self) -> bool {
        self.added_requirements.is_empty()
            && self.removed_requirements.is_empty()
            && self.modified_requirements.is_empty()
            && self.added_metrics.is_empty()
            && self.removed_metrics.is_empty()
            && self.added_risks.is_empty()
            && self.removed_risks.is_empty()
    }
}

/// `_by_id(requirements)` — Python dict comprehension semantics: on a
/// duplicate ID the LAST occurrence wins the value, but the key keeps its
/// FIRST-insertion position (dict key order).
fn by_id(requirements: &[Requirement]) -> Vec<(&str, &Requirement)> {
    let mut out: Vec<(&str, &Requirement)> = Vec::new();
    for r in requirements {
        if let Some(slot) = out.iter_mut().find(|(id, _)| *id == r.id) {
            slot.1 = r;
        } else {
            out.push((r.id.as_str(), r));
        }
    }
    out
}

/// `_ordered_difference(a, b)` — items in `a` not present in `b`, preserving
/// `a`'s order, de-duped.
fn ordered_difference(a: &[String], b: &[String]) -> Vec<String> {
    let mut seen: Vec<&str> = Vec::new();
    let mut out: Vec<String> = Vec::new();
    for item in a {
        if !b.contains(item) && !seen.contains(&item.as_str()) {
            seen.push(item.as_str());
            out.push(item.clone());
        }
    }
    out
}

/// `diff(old, new)` — the classified `Diff` between two products.
pub fn diff(old: &Artifact, new: &Artifact) -> Diff {
    let old_reqs = by_id(&old.product.requirements);
    let new_reqs = by_id(&new.product.requirements);

    let mut result = Diff::default();

    // Added / modified: iterate new (preserves new-file order).
    for (req_id, new_req) in &new_reqs {
        match old_reqs.iter().find(|(id, _)| id == req_id) {
            None => result.added_requirements.push((*new_req).clone()),
            Some((_, old_req)) if old_req.text != new_req.text => {
                result.modified_requirements.push(RequirementChange {
                    id: (*req_id).to_string(),
                    old_text: old_req.text.clone(),
                    new_text: new_req.text.clone(),
                });
            }
            Some(_) => {}
        }
    }

    // Removed: in old but not new (preserves old-file order).
    for (req_id, old_req) in &old_reqs {
        if !new_reqs.iter().any(|(id, _)| id == req_id) {
            result.removed_requirements.push((*old_req).clone());
        }
    }

    result.added_metrics =
        ordered_difference(&new.product.success_metrics, &old.product.success_metrics);
    result.removed_metrics =
        ordered_difference(&old.product.success_metrics, &new.product.success_metrics);
    result.added_risks = ordered_difference(&new.product.risks, &old.product.risks);
    result.removed_risks = ordered_difference(&old.product.risks, &new.product.risks);

    result
}
